### PR TITLE
change login failed message

### PR DESCRIPTION
**Pull request by lukyfebrianto**
**Description:** Improve edX login error message with reference to NEW UK region

**JIRA:** https://liv-it.atlassian.net/browse/WEB-2028

**Merge deadline:** List merge deadline (if any)

**Configuration instructions:** List any non-trivial configuration
instructions (if any).

**Testing instructions:**

1. Open page A
2. Do thing B
3. Expect C to happen
4. If D happened instead - check failed.

**Reviewers:**
- [x] tag reviewer
- [x] tag reviewer

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [x] Delete working branch (if not needed anymore)

**Author concerns:** List any concerns about this PR - inelegant
solutions, hacks, quick-and-dirty implementations, concerns about
migrations, etc.

### CODE REVIEW

File: common/djangoapps/student/views.py [hunk 1]
```diff
@@ -1236,9 +1236,21 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
                 AUDIT_LOG.warning(u"Login failed - password for user.id: {0} is invalid".format(loggable_id))
             else:
                 AUDIT_LOG.warning(u"Login failed - password for {0} is invalid".format(email))
+
+        # Labster set a new error message on UK region server
```
Comment (TamoshaytisV) [inline common/djangoapps/student/views.py]: I would rephrase this to 
```py
# Labster changes
# If there was login error, suggest student to login to UK region server
```
Or smth like this

File: lms/static/js/student_account/views/LoginView.js [hunk 2]
```diff
@@ -113,6 +113,11 @@
                 this.setErrors();
                 this.element.hide( this.$resetSuccess );
 
+                // Labster set a new error message title on UK region server
+                if (msg.toString().indexOf('https://courses-uk.labster.com')!=-1){
+                    this.$errors.find('.message-title').text('We could not sign you in. Wrong Region?')
```
Comment (TamoshaytisV) [inline lms/static/js/student_account/views/LoginView.js]: Isn't it better to change text here? https://github.com/Livit/Livit.Learn.EdX/blob/labster/develop/lms/templates/student_account/login.underscore#L20

**Review by TamoshaytisV — CHANGES_REQUESTED** [review]

File: lms/static/js/student_account/views/LoginView.js (hunk 2 above)
Comment (lukyfebrianto) [reply lms/static/js/student_account/views/LoginView.js]: I don't think so @TamoshaytisV ,
I make like you said in the previous commit but I realize when error login on UK region the message still same although the student on UK region, is that okay ?

File: common/djangoapps/student/views.py (hunk 1 above)
Comment (lukyfebrianto) [reply common/djangoapps/student/views.py]: changed

File: common/djangoapps/student/views.py [hunk 3]
```diff
@@ -1236,9 +1236,22 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
...
+
+        # Labster changes
+        # If there was login error, suggest student to login to UK region server
+        site_name = configuration_helpers.get_value('SITE_NAME', settings.SITE_NAME)
+        if not site_name == 'courses-uk.labster.com':
+            msg = _(
+                'If you are a student in the UK, your university may now be using the UK version of Labster. Please '
+                'try to log in here: <a href="https://courses-uk.labster.com">https://courses-uk.labster.com</a>'
+            )
+
```
Comment (TamoshaytisV) [inline common/djangoapps/student/views.py]: why do we need the empty line here?

File: common/djangoapps/student/views.py [hunk 4]
```diff
@@ -1236,9 +1236,22 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
                 AUDIT_LOG.warning(u"Login failed - password for user.id: {0} is invalid".format(loggable_id))
             else:
                 AUDIT_LOG.warning(u"Login failed - password for {0} is invalid".format(email))
+
+        # Labster changes
+        # If there was login error, suggest student to login to UK region server
+        site_name = configuration_helpers.get_value('SITE_NAME', settings.SITE_NAME)
+        if not site_name == 'courses-uk.labster.com':
```
Comment (TamoshaytisV) [inline common/djangoapps/student/views.py]: Can we place `courses-uk.labster.com` in lms/env/labster.py?
e.g.
```py
REGIONAL_EDX = 'courses-uk.labster.com'
```

**Review by TamoshaytisV — CHANGES_REQUESTED** [review]

**Review by TamoshaytisV — COMMENTED** [review]
> I make like you said in the previous commit but I realize when error login on UK region the message still same although the student on UK region, is that okay ?

I meant do site_name check again but this time in template
smth like
<% if not context.site_name == settings.REGIONAL_EDX %>
just make sure site_name is in the context

File: common/djangoapps/student/views.py (hunk 4 above)
Comment (polesye) [reply common/djangoapps/student/views.py]: As you might know, we might need to create a separate Regioanl servers for the students from other regions (like Canada etc.). 
And this solution will require additional development to support other regions too.
Can we simply use [CountryMiddleware](https://github.com/edx/edx-platform/blob/master/lms/envs/common.py#L1205) that adds [country code to the session](https://github.com/edx/edx-platform/blob/e8a36957b1f732974260e7b9b42b9c25148b492c/openedx/core/djangoapps/geoinfo/middleware.py#L44) ?

Chnages in `lms/envs/labster.py`
```py
REGIONS = {
  'uk': {'url': 'https://courses-uk.labster.com'},
}
```
In this file:
```py
regions = configuration_helpers.get_value('REGIONS', settings.REGIONS)
if request.session['country_code'] in regions:
  region_info = regions[request.session['country_code']]
  msg = _('...').format(region_url=region_info['url'])
else:
  msg = _('Email or password is incorrect.')
```
In this case, DevOps just need to add a new region info to the settings and it will work without additional code changes.
Note: Do not forget to make settings managable for the DevOps.

File: common/djangoapps/student/views.py [hunk 5]
```diff
@@ -1236,9 +1236,22 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
...
+                'try to log in here: <a href="https://courses-uk.labster.com">https://courses-uk.labster.com</a>'
+            )
+
+        else:
+            msg = _('Email or password is incorrect.')
+
         return JsonResponse({
             "success": False,
-            "value": _('Email or password is incorrect.'),
+            "value": msg,
```
Comment (polesye) [inline common/djangoapps/student/views.py:1260]: Who uses this message?

File: lms/static/js/student_account/views/LoginView.js [hunk 6]
```diff
@@ -113,6 +113,15 @@
                 this.setErrors();
                 this.element.hide( this.$resetSuccess );
 
+                /* Labster Changes
+                 * If there was login error, set title message to suggest
+                 * student login to UK region server
+                 */
+                var urlCheck = new RegExp("([a-zA-Z0-9]+://)?([a-zA-Z0-9_]+:[a-zA-Z0-9_]+@)?([a-zA-Z0-9.-]+\\.[A-Za-z]{2,4})(:[0-9]+)?([^ ])+");
+                if (msg.toString().match(urlCheck)){
+                    this.$errors.find('.message-title').text('We could not sign you in. Wrong Region?');
```
Comment (polesye) [inline lms/static/js/student_account/views/LoginView.js]: why do we need those changes? Are they so important?

File: common/djangoapps/student/views.py (hunk 5 above)
Comment (lukyfebrianto) [reply common/djangoapps/student/views.py:1260]: @polesye when student fail to sign in ( email or password is incorrect/doesn't match) will show this message, If student sign in to Edx central,

screenshot from 2017-09-12 17-38-35

If student sign in to Edx region,
screenshot from 2017-09-12 18-30-29

File: lms/static/js/student_account/views/LoginView.js (hunk 6 above)
Comment (lukyfebrianto) [reply lms/static/js/student_account/views/LoginView.js]: Yes its quite important when we want to change the title of error message
The code function is checking if message error contains an url, 
if contain url it means the message error referring to region error message and will change the title of message to 'We could not sign you in. Wrong region ?'

**Comment by lukyfebrianto** [comment]
> <% if not context.site_name == settings.REGIONAL_EDX %>
just make sure site_name is in the context

@TamoshaytisV, @polesye  do you know how to parse the value from python view to underscore file ? 
because the context from js file
https://github.com/Livit/Livit.Learn.EdX/blob/labster/develop/lms/static/js/student_account/views/LoginView.js#L43  

the other way is remove the title message on underscore file and we include the title message on failure response but we need to add/change the css file.

File: common/djangoapps/student/views.py (hunk 3 above)
Comment (lukyfebrianto) [reply common/djangoapps/student/views.py]: removed

File: common/djangoapps/student/views.py [hunk 7]
```diff
@@ -1236,9 +1236,19 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
                 AUDIT_LOG.warning(u"Login failed - password for user.id: {0} is invalid".format(loggable_id))
             else:
                 AUDIT_LOG.warning(u"Login failed - password for {0} is invalid".format(email))
+
+        # Labster changes
+        # If there was login error, suggest student to login to UK region server
+        if not request.build_absolute_uri('/')[:-1] == settings.LABSTER_REGIONAL_EDX:
+            msg = 'If you are a student in the UK, your university may now be using the UK version of Labster. Please' \
```
Comment (TamoshaytisV) [inline common/djangoapps/student/views.py]: Please change it to be more flexible
```py
regions = configuration_helpers.get_value('REGIONS', settings.REGIONS)
    if request.session['country_code'] in regions:
        region_info = regions[request.session['country_code']]
        msg = _(
            'If you are a student in {region_name}, your university may now be using the regional version of Labster. '
            'Please try to log in here: {region_login_url}'
        ).format(region_name=region_info['name'], region_login_url=region_info['login_url'])    
   else:
        msg = _('Email or password is incorrect.')
```

File: lms/envs/labster.py [hunk 8]
```diff
@@ -2,6 +2,7 @@
 import urlparse
 
 
+LABSTER_REGIONAL_EDX = 'https://courses-uk.labster.com'
```
Comment (TamoshaytisV) [inline lms/envs/labster.py]: Accept `REGIONS` using env tokens
Please provide full login link
```py
REGIONS = {
    'GB': {'login_url': 'https://courses-uk.labster.com/login', 'name': 'the United Kingdom'},
}
```

File: lms/static/js/student_account/views/LoginView.js [hunk 9]
```diff
@@ -113,6 +113,15 @@
                 this.setErrors();
                 this.element.hide( this.$resetSuccess );
 
+                /* Labster Changes
```
Comment (TamoshaytisV) [inline lms/static/js/student_account/views/LoginView.js]: Please remove changes in this file to prevent upgrade issues in future. It's ok to have default text `We can't sign you in`

**Review by TamoshaytisV — CHANGES_REQUESTED** [review]

File: lms/envs/labster.py [hunk 10]
```diff
@@ -2,7 +2,16 @@
 import urlparse
 
 
-LABSTER_REGIONAL_EDX = 'https://courses-uk.labster.com'
+# Regions of Labster
+DEFAULT_REGIONS = {
+  'uk': {
+      'url': 'https://courses-uk.labster.com',
+      'login_url': 'https://courses-uk.labster.com/login',
+      'name': 'United Kingdom'
```
Comment (polesye) [inline lms/envs/labster.py]: `United Kingdom` or `the United Kingdom`?

Comment (polesye) [reply lms/envs/labster.py]: Do we really need to fill this `DEFAULT_REGIONS` dict by default? Please ask DevOps to update servers-vers.yml to fill it instead.

File: common/djangoapps/student/views.py [hunk 11]
```diff
@@ -1238,11 +1238,20 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
...
+        if current_region in regions.keys():
+            region_info = regions[current_region]
+            msg = _(
+                'If you are a student in the {name}, your university may now be using the {region} version of '
+                'Labster. Please try to log in here: <a href="{login_url}">{site}</a>'
+            ).format(
+                name=region_info['name'],
+                region=current_region.upper(),
+                login_url=region_info['login_url'],
+                site=region_info['url']
```
Comment (polesye) [inline common/djangoapps/student/views.py]: Can we replace it with 'Please try to log in <a href="{login_url}">{region_name}</a> region' to avoid using redundant `region_info['url']`

File: common/djangoapps/student/views.py [hunk 12]
```diff
@@ -1238,11 +1238,20 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
...
-                  ' try to log in here: <a href="{site}{login_url}">{site}</a>'
-            msg = _(msg.format(site=settings.LABSTER_REGIONAL_EDX, login_url=reverse('signin_user')))
+        # If there was login error, suggest student to login to region server
+        regions = configuration_helpers.get_value('REGIONS', settings.REGIONS)
+        current_region = request.session.get('country_code')
+        if current_region in regions.keys():
+            region_info = regions[current_region]
+            msg = _(
+                'If you are a student in the {name}, your university may now be using the {region} version of '
+                'Labster. Please try to log in here: <a href="{login_url}">{site}</a>'
```
Comment (polesye) [inline common/djangoapps/student/views.py]: As a Translator I only see the text above and I know nothing about the code around and in which context it is used, so I can understand `{name}` placeholder as a Student Name, this Site name that is wrong and may affect the result translation. Please use `{region_name}` and `{region_ login_url}` instead.

File: lms/envs/labster.py [hunk 13]
```diff
@@ -2,6 +2,9 @@
 import urlparse
 
 
+# Regions of Labster
+REGIONS = ENV_TOKENS.get('LABSTER_REGIONS', {})
```
Comment (polesye) [inline lms/envs/labster.py:6]: @lukyfebrianto please create a tasl for DevOps with example of data that needds ot be added to the edx settings

File: common/djangoapps/student/views.py [hunk 14]
```diff
@@ -1236,9 +1236,27 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
                 AUDIT_LOG.warning(u"Login failed - password for user.id: {0} is invalid".format(loggable_id))
             else:
                 AUDIT_LOG.warning(u"Login failed - password for {0} is invalid".format(email))
+
+        # Labster changes
+        # If there was login error, suggest student to login to region server
+        regions = configuration_helpers.get_value('REGIONS', settings.REGIONS)
```
Comment (polesye) [inline common/djangoapps/student/views.py:1242]: `configuration_helpers` is an interesting feature that helps us manage settings directly from Dajngo Admin.
I think we should start using it to simplify changes in settings in the future. @sidorovdmitry what do you think?

screenshot 2017-09-13 09 12 11
screenshot 2017-09-13 09 12 28

Example
```json
"{\"REGIONS\": {\"GB\": {\"name\": \"the United Kingdom\", \"login_url\": \"https://staging-courses-uk.labster.com/dashboard\"}}}"
```

File: common/djangoapps/student/views.py [hunk 15]
```diff
@@ -1236,9 +1236,27 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
...
+        current_region = request.session.get('country_code')
```
Comment (polesye) [inline common/djangoapps/student/views.py]: @lukyfebrianto please leave a comment with a link to a list of Country Codes https://dev.maxmind.com/geoip/legacy/codes/iso3166/

File: common/djangoapps/student/views.py (hunk 14 above)
Comment (sydorovdmytro) [reply common/djangoapps/student/views.py:1242]: @polesye If it will help us, why not? 👍

**Comment by polesye** [comment]
@sidorovdmitry can you deploy this branch and setup the setting I posted to check how it works?

File: common/djangoapps/student/views.py (hunk 15 above)
Comment (lukyfebrianto) [reply common/djangoapps/student/views.py]: added

File: lms/envs/labster.py (hunk 13 above)
Comment (lukyfebrianto) [reply lms/envs/labster.py:6]: https://liv-it.atlassian.net/browse/OPS-523

**Comment by sydorovdmytro** [comment]
@polesye @lukyfebrianto done for https://edx-eucalyptus-sandbox.labster.com/
https://i.gyazo.com/5f6ee71d3061a69bff024c21de44ce34.png

**Review by polesye — APPROVED** [review]

**Review by TamoshaytisV — APPROVED** [review]
